File: packages/protocol/src/messages.ts
import { z } from "zod";
import type { SemanticSnapshot } from "./tree.js";
import type { LogRecord } from "./logs.js";
import type { ProbeInfo } from "./probe/ir.js";
import type { ProtocolLimits } from "./limits.js";
import { PROTOCOL_ID, type ProtocolId } from "./env.js";
import { ProtocolViolation } from "./errors.js";
import { projectDto } from "./framing.js";
import { validateSnapshot } from "./validate.js";
import { validateLogRecord } from "./logs.js";
import { probeInfoSchema, validateProbeInfo } from "./probe/validate.js";
import {
  ADAPTER_CAPABILITIES,
  EVIDENCE_PROVIDER_CAPABILITIES,
  type AdapterCapability,
  type EvidenceProviderRegistration,
} from "./contract.js";

/**
 * Wire messages. Transport: length-prefixed JSON frames (see framing.ts).
 * CDP-like: adapter pushes commits; driver issues requests; either side may
 * send errors. All messages are validated against limits BEFORE retention.
 */

export { ADAPTER_CAPABILITIES } from "./contract.js";
export type { AdapterCapability } from "./contract.js";

/** adapter → driver, exactly once, before any other message. */
export interface HelloMessage {
  readonly type: "hello";
  readonly protocol: ProtocolId;
  readonly token: string;
  readonly adapter: { readonly name: string; readonly version: string };
  readonly capabilities: readonly AdapterCapability[];
  /**
   * Present when the sender is a probe rather than a hand-written adapter.
   *
   * Carries what the probe can actually offer — framework and versions, the
   * best identity it can produce, and its optional abilities — so the driver
   * negotiates against measured capability rather than assuming a floor.
   */
  readonly probe?: ProbeInfo;
  /** Application evidence providers frozen into this session contract. */
  readonly providers?: readonly EvidenceProviderRegistration[];
}

/** driver → adapter, reply to hello. */
export interface HelloAckMessage {
  readonly type: "hello-ack";
  readonly protocol: ProtocolId;
  readonly sessionId: string;
  readonly limits: ProtocolLimits;
  /** Which semantic traffic the driver wants pushed. */
  readonly subscribe: "snapshots" | "revisions";
  /** Marker configuration: producer must emit the signed OSC 8487 commit marker. */
  readonly marker: { readonly enabled: boolean };
  /**
   * Log-channel budget, sent only when the adapter announced the `logs`
   * capability. **Absent means logs are disabled** — an adapter that receives
   * no `logs` field must not emit `log` messages at all.
   *
   * The adapter enforces the rate itself and drops locally when over budget,
   * leaving a gap in `LogRecord.seq` so the driver can report how many records
   * were lost. Enforcing it at the source is what keeps a log storm from
   * consuming the frame budget the semantic tree needs.
   */
  readonly logs?: {
    readonly enabled: boolean;
    /** Sustained ceiling on records per second. */
    readonly maxRecordsPerSecond: number;
    /** Records allowed in a burst on top of the sustained rate. */
    readonly burst: number;
  };
}

/** adapter → driver after each committed render (always, regardless of mode). */
export interface RevisionCommitMessage {
  readonly type: "revision-commit";
  readonly revision: number;
}

/** adapter → driver, full snapshot for a revision (subscribe: 'snapshots'). */
export interface SnapshotMessage {
  readonly type: "snapshot";
  readonly snapshot: SemanticSnapshot;
}

/**
 * adapter → driver, a frame has started (capability `frame-begin`).
 *
 * **Optional, and its absence means nothing.** No audited framework offers a
 * hook guaranteed to fire before every frame: one lets a pre-draw hook veto the
 * frame entirely, so the post-draw hook never runs; one exposes only a
 * post-frame hook; one decouples submission from the flush with a ticker. A
 * receiver that reads "no frame-begin" as "no frame in progress" turns four of
 * the six frameworks into a hang rather than an error.
 *
 * `FRAME_END` is the existing `revision-commit`, which stays advisory.
 *
 * **Abandoned frames**: a probe may begin a frame and never finish it — a
 * crash, an interrupted render. A `frame-begin` for revision N implicitly
 * closes every frame below N. Without that rule an open frame waits forever,
 * which is the timeout it replaced, only now wearing a false air of precision.
 */
export interface FrameBeginMessage {
  readonly type: "frame-begin";
  readonly revision: number;
}

/**
 * adapter → driver, one application log record (capability `logs`).
 *
 * Sent only after the driver enabled logs in `hello-ack`. Records are
 * independent of renders: they are not paired with a revision and never gate
 * snapshot publication.
 */
export interface LogMessage {
  readonly type: "log";
  readonly record: LogRecord;
}

/** either direction: terminal protocol error; sender closes after emitting. */
export interface ProtocolErrorMessage {
  readonly type: "error";
  readonly code:
    | "bad-token"
    | "bad-version"
    | "malformed"
    | "limit-exceeded"
    | "duplicate-semantic-key"
    | "adapter-guarantee-violation"
    | "capability-provider-violation"
    | "internal";
  readonly message: string;
}

export type AdapterToDriverMessage =
  | HelloMessage
  | RevisionCommitMessage
  | SnapshotMessage
  | FrameBeginMessage
  | LogMessage
  | ProtocolErrorMessage;

export type DriverToAdapterMessage = HelloAckMessage | ProtocolErrorMessage;

// --------------------------------------------------------------------------
// Runtime validation
//
// The interfaces above are the contract; the schemas below are how untrusted
// bytes become instances of it. Every parse projects the value into a frozen
// plain DTO first, so a getter on hostile input is rejected without running.
// --------------------------------------------------------------------------

/** Outcome of parsing one wire message. Mirrors `ProtocolErrorMessage['code']`. */
export type MessageParseResult<T> =
  | { readonly ok: true; readonly message: T }
  | {
      readonly ok: false;
      readonly code:
        | "bad-version"
        | "malformed"
        | "limit-exceeded"
        | "capability-provider-violation";
      readonly detail: string;
    };

/** Longest token/identifier/message string accepted, in UTF-16 code units. */
const MAX_IDENTIFIER_LENGTH = 1024;

const identifier = z.string().max(MAX_IDENTIFIER_LENGTH);
const nonEmptyIdentifier = identifier.min(1);
const safeIndex = z
  .number()
  .refine(
    (n) => Number.isSafeInteger(n) && n >= 0,
    "expected a non-negative safe integer",
  );
const revisionNumber = z
  .number()
  .refine(
    (n) => Number.isSafeInteger(n) && n > 0,
    "expected a positive safe integer",
  );

/**
 * Limits are an ADDITIVE part of the contract: unknown keys are IGNORED, not
 * rejected. A driver that learns a new ceiling must not break every already
 * published adapter, so this is the one object on the wire read leniently.
 * Known keys stay strict about their type, and every closed set elsewhere
 * (message types, roles, actions, capabilities) stays strict too.
 */
const limitsSchema = z.object({
  maxFrameBytes: revisionNumber,
  maxSnapshotBytes: revisionNumber,
  maxNodes: revisionNumber,
  maxDepth: revisionNumber,
  maxStringBytes: revisionNumber,
  maxRelationTargets: revisionNumber,
  maxQueuedFrames: revisionNumber,
  maxPendingWaiters: revisionNumber,
  maxSessions: revisionNumber,
  maxLogRecordBytes: revisionNumber,
  maxLogQueue: revisionNumber,
});

const errorFields = {
  type: z.literal("error"),
  code: z.enum([
    "bad-token",
    "bad-version",
    "malformed",
    "limit-exceeded",
    "duplicate-semantic-key",
    "adapter-guarantee-violation",
    "capability-provider-violation",
    "internal",
  ]),
  message: z.string().max(MAX_IDENTIFIER_LENGTH),
};

/** adapter → driver: strict, this is the hostile-input boundary. */
const errorSchema = z.strictObject(errorFields);

/** driver → adapter: tolerant envelope, see the note above `parseDriverMessage`. */
const errorFromDriverSchema = z.object(errorFields);

/** adapter → driver schemas. Snapshot bodies are validated separately. */
const helloSchema = z.strictObject({
  type: z.literal("hello"),
  protocol: z.literal(PROTOCOL_ID),
  token: nonEmptyIdentifier,
  adapter: z.strictObject({
    name: nonEmptyIdentifier,
    version: nonEmptyIdentifier,
  }),
  capabilities: z
    .array(z.enum(ADAPTER_CAPABILITIES))
    .max(ADAPTER_CAPABILITIES.length),
  probe: probeInfoSchema.optional(),
  providers: z
    .array(
      z.strictObject({
        id: nonEmptyIdentifier,
        version: nonEmptyIdentifier,
        method: z.enum(["native", "declared"]),
        capabilities: z
          .array(z.enum(EVIDENCE_PROVIDER_CAPABILITIES))
          .min(1)
          .max(EVIDENCE_PROVIDER_CAPABILITIES.length)
          .refine(
            (values) => new Set(values).size === values.length,
            "provider capabilities must be unique",
          ),
      }),
    )
    .max(64)
    .superRefine((providers, ctx) => {
      const ids = new Set<string>();
      for (let index = 0; index < providers.length; index += 1) {
        const provider = providers[index]!;
        const id = provider.id;
        if (ids.has(id)) {
          ctx.addIssue({
            code: "custom",
            path: [index, "id"],
            message: `duplicate provider id ${id}`,
          });
          return;
        }
        ids.add(id);
      }
    })
    .optional(),
});

const frameBeginSchema = z.strictObject({
  type: z.literal("frame-begin"),
  revision: revisionNumber,
});

const revisionCommitSchema = z.strictObject({
  type: z.literal("revision-commit"),
  revision: revisionNumber,
});

const snapshotEnvelopeSchema = z.strictObject({
  type: z.literal("snapshot"),
  snapshot: z.unknown(),
});

const logEnvelopeSchema = z.strictObject({
  type: z.literal("log"),
  record: z.unknown(),
});

/** driver → adapter schemas. */
const helloAckSchema = z.object({
  type: z.literal("hello-ack"),
  protocol: z.literal(PROTOCOL_ID),
  sessionId: nonEmptyIdentifier,
  limits: limitsSchema,
  subscribe: z.enum(["snapshots", "revisions"]),
  marker: z.object({ enabled: z.boolean() }),
  logs: z
    .object({
      enabled: z.boolean(),
      maxRecordsPerSecond: revisionNumber,
      burst: safeIndex,
    })
    .optional(),
});

function malformed(detail: string): MessageParseResult<never> {
  return { ok: false, code: "malformed", detail };
}

/** Projection guard shared by both parsers. */
function project(
  value: unknown,
  limits: ProtocolLimits,
): MessageParseResult<unknown> {
  try {
    return { ok: true, message: projectDto<unknown>(value, limits.maxDepth) };
  } catch (error) {
    const detail =
      error instanceof ProtocolViolation
        ? error.message
        : "value is not a plain JSON DTO";
    return error instanceof ProtocolViolation && error.code === "dto-depth"
      ? { ok: false, code: "limit-exceeded", detail }
      : malformed(detail);
  }
}

function messageType(value: unknown): string | null {
  if (typeof value !== "object" || value === null) return null;
  const type: unknown = (value as { type?: unknown }).type;
  return typeof type === "string" ? type : null;
}

function check(schema: z.ZodType, value: unknown): string | null {
  const result = schema.safeParse(value);
  if (result.success) return null;
  const issue = result.error.issues[0]!;
  const where =
    issue.path.length > 0 ? issue.path.map(String).join(".") : "<root>";
  return `${where}: ${issue.message}`;
}

/**
 * Validate a snapshot carried inside an envelope and map its failure onto the
 * wire error taxonomy: capacity failures are `limit-exceeded`; evidence-frame
 * contract failures retain their typed provider classification; the rest are
 * `malformed`.
 */
function checkSnapshot(
  value: unknown,
  limits: ProtocolLimits,
): MessageParseResult<never> | null {
  const result = validateSnapshot(value, limits);
  if (result.ok) return null;
  const overCapacity =
    result.code === "bytes" ||
    result.code === "count" ||
    result.code === "depth" ||
    result.code === "string-bytes";
  return {
    ok: false,
    code: overCapacity
      ? "limit-exceeded"
      : result.code === "provider"
        ? "capability-provider-violation"
        : "malformed",
    detail: `snapshot ${result.code}: ${result.detail}`,
  };
}

/**
 * Validate a log record carried inside an envelope, mapping capacity failures
 * onto `limit-exceeded` exactly as snapshots do.
 */
function checkLogRecord(
  value: unknown,
  limits: ProtocolLimits,
): MessageParseResult<never> | null {
  const result = validateLogRecord(value, limits);
  if (result.ok) return null;
  const overCapacity =
    result.code === "bytes" ||
    result.code === "count" ||
    result.code === "depth" ||
    result.code === "string-bytes";
  return {
    ok: false,
    code: overCapacity ? "limit-exceeded" : "malformed",
    detail: `log record ${result.code}: ${result.detail}`,
  };
}

/**
 * Parse and validate one adapter → driver message.
 *
 * **Strict reader**: this is the hostile-input boundary, so unknown fields are
 * rejected rather than ignored. See {@link parseDriverMessage} for why the
 * other direction is tolerant.
 *
 * @param value - Untrusted decoded frame body.
 * @param limits - Active session limits, applied to any embedded snapshot.
 * @returns A frozen message on success, or a typed failure. Never throws.
 */
export function parseAdapterMessage(
  value: unknown,
  limits: ProtocolLimits,
): MessageParseResult<AdapterToDriverMessage> {
  const projected = project(value, limits);
  if (!projected.ok) return projected;
  const dto = projected.message;

  switch (messageType(dto)) {
    case "hello": {
      const protocol: unknown = (dto as { protocol?: unknown }).protocol;
      if (typeof protocol === "string" && protocol !== PROTOCOL_ID) {
        return {
          ok: false,
          code: "bad-version",
          detail: `unsupported protocol ${protocol}`,
        };
      }
      const issue = check(helloSchema, dto);
      if (issue !== null) return malformed(issue);
      // The shape check cannot see the one incoherent pair: a probe declaring
      // frame-local identity while claiming it can be correlated across
      // frames. That rule has to hold on the wire, not only when a caller
      // remembers to run the helper.
      const probe = (dto as { probe?: unknown }).probe;
      if (probe !== undefined) {
        const checked = validateProbeInfo(probe);
        if (!checked.ok) return malformed(`probe: ${checked.detail}`);
      }
      return { ok: true, message: dto as HelloMessage };
    }
    case "revision-commit": {
      const issue = check(revisionCommitSchema, dto);
      return issue === null
        ? { ok: true, message: dto as RevisionCommitMessage }
        : malformed(issue);
    }
    case "snapshot": {
      const issue = check(snapshotEnvelopeSchema, dto);
      if (issue !== null) return malformed(issue);
      const bad = checkSnapshot(
        (dto as { snapshot: unknown }).snapshot,
        limits,
      );
      return bad ?? { ok: true, message: dto as SnapshotMessage };
    }
    case "frame-begin": {
      const issue = check(frameBeginSchema, dto);
      return issue === null
        ? { ok: true, message: dto as FrameBeginMessage }
        : malformed(issue);
    }
    case "log": {
      const issue = check(logEnvelopeSchema, dto);
      if (issue !== null) return malformed(issue);
      const bad = checkLogRecord((dto as { record: unknown }).record, limits);
      return bad ?? { ok: true, message: dto as LogMessage };
    }
    case "error": {
      const issue = check(errorSchema, dto);
      return issue === null
        ? { ok: true, message: dto as ProtocolErrorMessage }
        : malformed(issue);
    }
    default:
      return malformed("unknown or missing message type");
  }
}

/**
 * Parse and validate one driver → adapter message.
 *
 * **Tolerant reader.** Unlike {@link parseAdapterMessage}, unknown envelope
 * fields are ignored rather than rejected, and are carried through to the
 * caller so a reader that does understand them still can. Known fields stay
 * strictly type-checked, and closed sets (`type`, `code`, `subscribe`) stay
 * closed — an unknown message type is still `malformed`.
 *
 * The asymmetry is about who is speaking, not about the message. The driver is
 * the trusted party and behaviour is governed by negotiated capabilities, so a
 * newer driver may add an optional field without invalidating every adapter
 * already published. Traffic in the other direction crosses the hostile-input
 * boundary and stays strict.
 *
 * @param value - Decoded frame body from the driver.
 * @param limits - Active session limits used for the projection depth bound.
 * @returns A frozen message on success, or a typed failure. Never throws.
 */
export function parseDriverMessage(
  value: unknown,
  limits: ProtocolLimits,
): MessageParseResult<DriverToAdapterMessage> {
  const projected = project(value, limits);
  if (!projected.ok) return projected;
  const dto = projected.message;

  switch (messageType(dto)) {
    case "hello-ack": {
      const protocol: unknown = (dto as { protocol?: unknown }).protocol;
      if (typeof protocol === "string" && protocol !== PROTOCOL_ID) {
        return {
          ok: false,
          code: "bad-version",
          detail: `unsupported protocol ${protocol}`,
        };
      }
      const issue = check(helloAckSchema, dto);
      return issue === null
        ? { ok: true, message: dto as HelloAckMessage }
        : malformed(issue);
    }
    case "error": {
      const issue = check(errorFromDriverSchema, dto);
      return issue === null
        ? { ok: true, message: dto as ProtocolErrorMessage }
        : malformed(issue);
    }
    default:
      return malformed("unknown or missing message type");
  }
}

File: clients/go/protocol/messages.go
package protocol

import (
	"fmt"
)

// ProtocolID is the wire protocol identifier both sides must agree on.
const ProtocolID = "termwright/3"

// ProtocolVersion is the current major version.
const ProtocolVersion = 3

// maxIdentifierLength bounds tokens, ids and free-text error messages.
const maxIdentifierLength = 1024

// Hello is the adapter's handshake: sent exactly once, before anything else.
type Hello struct {
	Type         string                         `json:"type"`
	Protocol     string                         `json:"protocol"`
	Token        string                         `json:"token"`
	Adapter      AdapterInfo                    `json:"adapter"`
	Capabilities []Capability                   `json:"capabilities"`
	Probe        *ProbeInfo                     `json:"probe,omitempty"`
	Providers    []EvidenceProviderRegistration `json:"providers,omitempty"`
}

// EvidenceProviderRegistration freezes an application evidence producer in
// the same hello negotiation as the framework adapter.
type EvidenceProviderRegistration struct {
	ID           string   `json:"id"`
	Version      string   `json:"version"`
	Method       string   `json:"method"` // native or declared
	Capabilities []string `json:"capabilities"`
}

// EvidenceProviderRegistry freezes application providers once per client session.
type EvidenceProviderRegistry interface {
	Freeze() (EvidenceProviderLease, error)
}

// EvidenceProviderLease is the immutable provider set owned by one session.
type EvidenceProviderLease interface {
	Registrations() []EvidenceProviderRegistration
	Collect(sessionID string, revision int64, columns, rows int) []ProviderRevisionEvidence
	Close()
}

// AdapterInfo identifies the adapter implementation to the driver.
type AdapterInfo struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}

// HelloAck is the driver's reply: the session id, the negotiated limits, what
// to push, and whether markers are wanted.
type HelloAck struct {
	Type      string       `json:"type"`
	Protocol  string       `json:"protocol"`
	SessionID string       `json:"sessionId"`
	Limits    Limits       `json:"limits"`
	Subscribe string       `json:"subscribe"`
	Marker    MarkerConfig `json:"marker"`
	Logs      *LogBudget   `json:"logs,omitempty"`
}

// MarkerConfig says whether the adapter should emit render markers.
type MarkerConfig struct {
	Enabled bool `json:"enabled"`
}

// LogBudget is the log-channel allowance, sent only when the adapter
// announced the `logs` capability. Absent means logs are disabled: an adapter
// that receives no budget must not emit log messages at all.
type LogBudget struct {
	Enabled bool `json:"enabled"`
	// MaxRecordsPerSecond is the sustained ceiling on records per second.
	MaxRecordsPerSecond int `json:"maxRecordsPerSecond"`
	// Burst is how many records are allowed on top of the sustained rate.
	Burst int `json:"burst"`
}

// RevisionCommit announces that a render was committed to the terminal.
type RevisionCommit struct {
	Type     string `json:"type"`
	Revision int64  `json:"revision"`
}

// SemanticFullMessage carries a full tree for one revision.
type SemanticFullMessage struct {
	Type     string    `json:"type"`
	Snapshot *Snapshot `json:"snapshot"`
}

// ProtocolErrorMessage is terminal: the sender closes after emitting it.
type ProtocolErrorMessage struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

// NewHello builds a handshake message, refusing unknown capabilities locally.
func NewHello(token, name, version string, capabilities []Capability) (*Hello, error) {
	return newHello(token, name, version, capabilities, nil)
}

func newHello(token, name, version string, capabilities []Capability, probe *ProbeInfo) (*Hello, error) {
	for _, capability := range capabilities {
		if !ValidCapability(capability) {
			return nil, violation("marker-argument", "unknown capability %q", capability)
		}
	}
	checkedProbe, err := checkedProbeInfo(probe)
	if err != nil {
		return nil, violation("marker-argument", "invalid probe info: %v", err)
	}
	return &Hello{
		Type:         "hello",
		Protocol:     ProtocolID,
		Token:        token,
		Adapter:      AdapterInfo{Name: name, Version: version},
		Capabilities: capabilities,
		Probe:        checkedProbe,
	}, nil
}

// ParseError reports why a message was refused, in the wire taxonomy:
// bad-version, malformed or limit-exceeded.
type ParseError struct {
	Code   string
	Detail string
}

func (e *ParseError) Error() string { return e.Code + ": " + e.Detail }

// ParseCode returns the code of a *ParseError, or "" otherwise.
func ParseCode(err error) string {
	if e, ok := err.(*ParseError); ok {
		return e.Code
	}
	return ""
}

func malformed(format string, args ...any) *ParseError {
	return &ParseError{Code: "malformed", Detail: fmt.Sprintf(format, args...)}
}

var errorCodes = map[string]struct{}{
	"bad-token": {}, "bad-version": {}, "malformed": {}, "limit-exceeded": {}, "duplicate-semantic-key": {}, "adapter-guarantee-violation": {}, "internal": {},
}

func project(value any, limits Limits) (any, *ParseError) {
	projected, err := ProjectDTO(value, limits.MaxDepth)
	if err != nil {
		if ViolationCode(err) == "dto-depth" {
			return nil, &ParseError{Code: "limit-exceeded", Detail: err.Error()}
		}
		return nil, malformed("%s", err.Error())
	}
	return projected, nil
}

// requiredKeys checks that every required key is present, tolerating unknown ones.
func requiredKeys(object map[string]any, required []string) *ParseError {
	for _, key := range required {
		if _, ok := object[key]; !ok {
			return malformed("missing field %q", key)
		}
	}
	return nil
}

func requireKeys(object map[string]any, required []string, optional []string) *ParseError {
	for _, key := range required {
		if _, ok := object[key]; !ok {
			return malformed("missing field %q", key)
		}
	}
	for key := range object {
		known := false
		for _, name := range append(append([]string{}, required...), optional...) {
			if key == name {
				known = true
				break
			}
		}
		if !known {
			return malformed("unrecognized key %q", key)
		}
	}
	return nil
}

func identifier(object map[string]any, key string, allowEmpty bool) *ParseError {
	text, ok := object[key].(string)
	if !ok {
		return malformed("%s: expected a string", key)
	}
	if len(text) > maxIdentifierLength {
		return malformed("%s: expected at most %d characters", key, maxIdentifierLength)
	}
	if !allowEmpty && text == "" {
		return malformed("%s: expected a non-empty string", key)
	}
	return nil
}

func wholeNumber(object map[string]any, key string, positive bool) *ParseError {
	number, ok := object[key].(float64)
	if !ok || number != float64(int64(number)) || number > maxSafeInteger {
		return malformed("%s: expected a safe integer", key)
	}
	if positive && number <= 0 {
		return malformed("%s: expected a positive safe integer", key)
	}
	if !positive && number < 0 {
		return malformed("%s: expected a non-negative safe integer", key)
	}
	return nil
}

func checkEmbeddedSnapshot(value any, limits Limits) *ParseError {
	err := ValidateSnapshot(value, limits)
	if err == nil {
		return nil
	}
	code := ValidationCode(err)
	wire := "malformed"
	switch code {
	case "bytes", "count", "depth", "string-bytes":
		wire = "limit-exceeded"
	}
	return &ParseError{Code: wire, Detail: "snapshot " + err.Error()}
}

// checkLogBudget validates the optional log-channel budget in hello-ack.
func checkLogBudget(value any) *ParseError {
	budget, ok := value.(map[string]any)
	if !ok {
		return malformed("logs: expected an object")
	}
	if problem := requiredKeys(budget, []string{"enabled", "maxRecordsPerSecond", "burst"}); problem != nil {
		return problem
	}
	if _, ok := budget["enabled"].(bool); !ok {
		return malformed("logs.enabled: expected a boolean")
	}
	if problem := wholeNumber(budget, "maxRecordsPerSecond", true); problem != nil {
		return problem
	}
	return wholeNumber(budget, "burst", false)
}

func checkErrorMessage(object map[string]any, strict bool) *ParseError {
	keys := []string{"type", "code", "message"}
	problem := requiredKeys(object, keys)
	if strict {
		problem = requireKeys(object, keys, nil)
	}
	if problem != nil {
		return problem
	}
	code, _ := object["code"].(string)
	if _, known := errorCodes[code]; !known {
		return malformed("code: unknown error code")
	}
	return identifier(object, "message", true)
}

func messageObject(value any) (map[string]any, string, *ParseError) {
	object, ok := value.(map[string]any)
	if !ok {
		return nil, "", malformed("unknown or missing message type")
	}
	kind, ok := object["type"].(string)
	if !ok {
		return nil, "", malformed("unknown or missing message type")
	}
	return object, kind, nil
}

func checkProtocolField(object map[string]any) *ParseError {
	protocol, ok := object["protocol"].(string)
	if !ok {
		return malformed("protocol: expected a string")
	}
	if protocol != ProtocolID {
		return &ParseError{Code: "bad-version", Detail: "unsupported protocol " + protocol}
	}
	return nil
}

// ParseAdapterMessage validates one adapter → driver message and returns the
// checked generic form. It never panics; failures come back as *ParseError.
//
// Strict: an unknown field from an adapter is a protocol error, not an
// extension. See ParseDriverMessage for the other direction.
func ParseAdapterMessage(value any, limits Limits) (map[string]any, error) {
	projected, parseErr := project(value, limits)
	if parseErr != nil {
		return nil, parseErr
	}
	object, kind, parseErr := messageObject(projected)
	if parseErr != nil {
		return nil, parseErr
	}

	switch kind {
	case "hello":
		if problem := checkProtocolField(object); problem != nil {
			return nil, problem
		}
		if problem := requireKeys(object, []string{"type", "protocol", "token", "adapter", "capabilities"}, []string{"probe"}); problem != nil {
			return nil, problem
		}
		if problem := identifier(object, "token", false); problem != nil {
			return nil, problem
		}
		adapter, ok := object["adapter"].(map[string]any)
		if !ok {
			return nil, malformed("adapter: expected an object")
		}
		if problem := requireKeys(adapter, []string{"name", "version"}, nil); problem != nil {
			return nil, problem
		}
		for _, key := range []string{"name", "version"} {
			if problem := identifier(adapter, key, false); problem != nil {
				return nil, problem
			}
		}
		capabilities, ok := object["capabilities"].([]any)
		if !ok || len(capabilities) > CapabilityCount {
			return nil, malformed("capabilities: expected a bounded array")
		}
		for _, item := range capabilities {
			name, isString := item.(string)
			if !isString || !ValidCapability(Capability(name)) {
				return nil, malformed("capabilities: unknown capability")
			}
		}
		if probe, present := object["probe"]; present {
			if problem := checkProbeInfo(probe); problem != nil {
				return nil, problem
			}
		}
		return object, nil

	case "revision-commit":
		if problem := requireKeys(object, []string{"type", "revision"}, nil); problem != nil {
			return nil, problem
		}
		if problem := wholeNumber(object, "revision", true); problem != nil {
			return nil, problem
		}
		return object, nil

	case "semantic-full":
		if problem := requireKeys(object, []string{"type", "snapshot"}, nil); problem != nil {
			return nil, problem
		}
		if problem := checkEmbeddedSnapshot(object["snapshot"], limits); problem != nil {
			return nil, problem
		}
		return object, nil

	case "log":
		if problem := requireKeys(object, []string{"type", "record"}, nil); problem != nil {
			return nil, problem
		}
		if problem := checkLogRecord(object["record"], limits); problem != nil {
			return nil, problem
		}
		return object, nil

	case "error":
		if problem := checkErrorMessage(object, true); problem != nil {
			return nil, problem
		}
		return object, nil
	}
	return nil, malformed("unknown or missing message type")
}

// checkLogRecord maps a record failure onto the wire taxonomy: capacity
// failures are limit-exceeded, the rest are malformed.
func checkLogRecord(value any, limits Limits) *ParseError {
	err := ValidateLogRecord(value, limits)
	if err == nil {
		return nil
	}
	wire := "malformed"
	switch ValidationCode(err) {
	case "bytes", "count", "depth", "string-bytes":
		wire = "limit-exceeded"
	}
	return &ParseError{Code: wire, Detail: "log record " + err.Error()}
}

// ParseDriverMessage validates one driver → adapter message.
//
// Driver traffic is read tolerantly: unknown fields in the envelope and in the
// driver's nested objects (marker, logs, limits) are ignored and passed
// through to the caller, so a newer driver can add a field without breaking an
// adapter published before it existed.
//
// The asymmetry is about who is speaking, not about the message: adapter
// traffic crosses an untrusted boundary, where an unknown field is a signal
// rather than an extension. Tolerance is not leniency either — known fields
// keep their types, and the closed sets (message types, error codes,
// subscribe, roles, actions) stay closed in both directions.
func ParseDriverMessage(value any, limits Limits) (map[string]any, error) {
	projected, parseErr := project(value, limits)
	if parseErr != nil {
		return nil, parseErr
	}
	object, kind, parseErr := messageObject(projected)
	if parseErr != nil {
		return nil, parseErr
	}

	switch kind {
	case "hello-ack":
		if problem := checkProtocolField(object); problem != nil {
			return nil, problem
		}
		if problem := requiredKeys(object,
			[]string{"type", "protocol", "sessionId", "limits", "subscribe", "marker"}); problem != nil {
			return nil, problem
		}
		if problem := identifier(object, "sessionId", false); problem != nil {
			return nil, problem
		}
		limitsObject, ok := object["limits"].(map[string]any)
		if !ok {
			return nil, malformed("limits: expected an object")
		}
		// Required keys must all be present, but unknown ones are ignored:
		// `limits` is the one object on the wire that grows between versions,
		// and a client that rejected a ceiling it had never heard of would
		// drop the channel every time the protocol gained one.
		if problem := requiredKeys(limitsObject, limitFields); problem != nil {
			return nil, problem
		}
		for _, key := range limitFields {
			if problem := wholeNumber(limitsObject, key, true); problem != nil {
				return nil, problem
			}
		}
		subscribe, _ := object["subscribe"].(string)
		if subscribe != "semantic" {
			return nil, malformed("subscribe: expected 'semantic'")
		}
		marker, ok := object["marker"].(map[string]any)
		if !ok {
			return nil, malformed("marker: expected an object")
		}
		if problem := requiredKeys(marker, []string{"enabled"}); problem != nil {
			return nil, problem
		}
		if _, ok := marker["enabled"].(bool); !ok {
			return nil, malformed("marker.enabled: expected a boolean")
		}
		if logs, present := object["logs"]; present {
			if problem := checkLogBudget(logs); problem != nil {
				return nil, problem
			}
		}
		return object, nil

	case "semantic-resync-request":
		if problem := requiredKeys(object, []string{"type", "sessionId", "expectedBaseRevision", "reason"}); problem != nil {
			return nil, problem
		}
		if problem := identifier(object, "sessionId", false); problem != nil {
			return nil, problem
		}
		if object["expectedBaseRevision"] != nil {
			if problem := wholeNumber(object, "expectedBaseRevision", true); problem != nil {
				return nil, problem
			}
		}
		reason, _ := object["reason"].(string)
		if reason != "base-mismatch" && reason != "missing-base" && reason != "driver-reset" {
			return nil, malformed("reason: unknown semantic resync reason")
		}
		return object, nil

	case "error":
		if problem := checkErrorMessage(object, false); problem != nil {
			return nil, problem
		}
		return object, nil
	}
	return nil, malformed("unknown or missing message type")
}

var limitFields = []string{
	"maxFrameBytes", "maxSnapshotBytes", "maxNodes", "maxDepth", "maxStringBytes",
	"maxRelationTargets", "maxQueuedFrames", "maxPendingWaiters", "maxSessions",
	"maxLogRecordBytes", "maxLogQueue",
}
